app/models/spree/image.rb

```diff
@@ -25,6 +25,7 @@ def variant(name)
 
     def url(size)
       return unless attachment.attached?
+      return variant(size).processed.url if attachment.service.name == :amazon_public
```

Comment:
Oh, I hoped that this wouldn't be necessary when using a public bucket. Hm, possible to do this only when the variant has been processed already? Otherwise it's a blocking call and it could make a request really slow.

Reply:
With named variants they get created immediately when the image is first uploaded instead of lazily afterwards on demand, so any variants should be generally be available and processed (in theory) after that initial upload request is finished. But yeah I think this is necessary if we're rendering an image src that directly points to S3. 

It's also another reason to use `Vips` for image processing, as it's 18 times faster :)

Reply:
Ah, okay, so the window of unprocessed variants is very small. The `processed?` method is private. So we would need to call `service.exist?(key)` ourselves. I would probably do that here but I'm also okay to avoid that complexity.

Reply:
I think if there's ever a situation where `processed` gets called and it isn't finished yet it blocks until it's finished. It should always return the processed URL though. I think it's not worth asking `service.exist?(key)` every time here, as that will make an additional API call to S3, and 99.99% of the time it'll be processed already at the point we're looking for it.

Reply:
Yes, `processed` just blocks and it always returns the processed URL. And I also agree with not calling `exist?` for every image. Unfortunately, I just found that `processed` is calling it already. So on a page with 25 products, we make 25 S3 API calls to find that the variants are already processed. I thought that Rails fixed this by storing variant records...

https://github.com/rails/rails/blob/55da913c4158804134ba73e03489c6c8125ef05d/activestorage/app/models/active_storage/variant.rb#L65-L68

https://github.com/rails/rails/blob/55da913c4158804134ba73e03489c6c8125ef05d/activestorage/app/models/active_storage/variant.rb#L111-L113

If that's true, we have a trade-off here. 25 external requests in one app request may still be better than 25 app requests with one external request each. :shrug:

Reply:
Ah I see. Yes I think it would still get called at some point if it was going through the `/active_storage/representations/redirect` route. :man_shrugging: